Service/Runner.php

```diff
@@ -189,40 +162,70 @@ private function shouldStoreState($componentId, $configurationId)
     }
 
     /**
-     * @param array $componentData
-     * @param $configId
-     * @param array $configData
-     * @param array $state
-     * @param string $action
-     * @param string $mode
-     * @param string $jobId
-     * @return Output
+     * @param JobDefinition $jobDefinition
+     * @param $action
+     * @param $mode
+     * @param $jobId
+     * @param Output $output
      */
-    public function run(array $componentData, $configId, array $configData, array $state, $action, $mode, $jobId)
+    public function runRow(JobDefinition $jobDefinition, $action, $mode, $jobId, Output &$output)
```

Comment:
potrebujem & ?

Reply:
dobrá otázka, ocargokultil jsem to z jinýho místa. kdyby to šlo nějak jinak, bylo by to super. v jednotlivejch rows/procesorech se tam furt přidávaj images...

Reply:
tak s tím outputem ještě čuju nějakej zádrhel. 

```
        $componentOutput = new Output();
        foreach ($jobDefinitions as $jobDefinition) {
            if ($jobDefinition->isDisabled()) {
                $this->loggerService->getLog()->notice(
                    "Skipping configuration id: " . $jobDefinition->getConfigId() . ' version:' . $jobDefinition->getConfigVersion()
                    . ", rowId: " . $jobDefinition->getRowId()
                );
                continue;
            }
            $this->runRow($jobDefinition, $action, $mode, $jobId, $componentOutput);
        }
```

Jenže `Output` classa neumí appendovat (dle názvu `Output::addProcessOutput()` jsem myslel, že to bude append, ale on je to spíš setter), stejně tak `Output::addImages()`. Tohle může fungovat pro 1 row, ale pro víc rows je to totální zmatek. `Runner::run()` bude muset vracet `array Output[]`, přejmenuju ty metody z `add*()` na `set*()` a pak se to nebude muset předávat argumentem do metod, ale jenom to probuble ven z `Runner::runImages()`, kde se to stvoří. 

No hlavně na to budu muset napsat testy, aby odchytily to současný vadný chování a pak to fixnout.

Reply:
Udělal jsem nakonec z `Output`u immutable. Jedinej kus, kterej tam zbyl a kterýmu nerozumím, je

```
                if ($image->isMain()) {
                    $outputMessage = $process->getOutput();
                }
```
(https://github.com/keboola/docker-bundle/blob/0b341c3446d4c2a0daa4ded49512ad2661b90d6a/Service/Runner.php#L445)

Metoda `Output::getProcessOutput` se používá jen v `ActionController` (https://github.com/keboola/docker-bundle/blob/0b341c3446d4c2a0daa4ded49512ad2661b90d6a/Controller/ActionController.php#L91) a tam by se žádný procesory nebo jiný image pouštět neměly, ne? Takže bych mohl tu podmínku vyhodit a nasypat to do všech `Output`ů. Na konci se stejně agregují jen digesty.

Reply:
Ne, jasně, tu podmínku `isMain()` už mám, to je na pipeline, chjo, se mi to pomotalo. Ale teda napadá mě, jestli tam neukládat výstup všech images.

Reply:
jeste jsem to nezpracoval, jen jsem chtel rict, ze & by tam byl potreba jen kdyz by se menil pointer na ten objekt, ze se meni neco na tom objektu nevadi, ale jeste se podivam

Reply:
no rekli jsme, ze na vystup procesoru prdime, a drzel bych se toho
ale bude potreba ulozit vsechny images - ted sice budou vsechny stejny, ale az bude multicomponent tak nebudou

Reply:
co je multicomponent?

Reply:
https://github.com/keboola/docker-bundle/issues/175

Reply:
@odin veškerý `&` jsem vyhodil, je to immutable a probublává to nahoru.